### ISSUE TITLE
Not working on macOS 10.13.6

### ISSUE THREAD

**Issue by richardsimko** [hasse69/rar2fs#138]
I had a working installation of rar2fs 1.26.0 but decided to update both that and osxfuse. Unfortunately this has now completely broken the installation. I have issues similar to #119 but I've tried all the troubleshooting there to no avail.

When I mount a file (`rar2fs -d ~/Downloads/test/example.rar ~/Downloads/test/mnt -o volname=test`) the directory I'm targeting disappears from Finder and does not appear as an external drive as it used to. If I try to access it through the terminal I get an IO error. The output from rar2fs does not really point to anything obvious:

```                                      
Apr 13 14:20:10  rar2fs[48662] <Debug>: mounted /Users/user/Downloads/test/mnt
unique: 2, opcode: INIT (26), nodeid: 0, insize: 56
INIT: 7.19
flags=0xf8000008
max_readahead=0x00100000
   INIT: 7.8
   flags=0x00000000
   max_readahead=0x00100000
   max_write=0x02000000
   unique: 2, error: 0 (Undefined error: 0), outsize: 40
unique: 2, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 2, error: 0 (Undefined error: 0), outsize: 96
unique: 2, opcode: GETATTR (3), nodeid: 1, insize: 40
   unique: 2, error: 0 (Undefined error: 0), outsize: 128
unique: 2, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 2, error: 0 (Undefined error: 0), outsize: 96
unique: 3, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 3, error: 0 (Undefined error: 0), outsize: 96
unique: 2, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 2, error: 0 (Undefined error: 0), outsize: 96
unique: 2, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 2, error: 0 (Undefined error: 0), outsize: 96
unique: 4, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 4, error: 0 (Undefined error: 0), outsize: 96
unique: 5, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 5, error: 0 (Undefined error: 0), outsize: 96
unique: 6, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 6, error: 0 (Undefined error: 0), outsize: 96
unique: 7, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 7, error: 0 (Undefined error: 0), outsize: 96
unique: 8, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 8, error: 0 (Undefined error: 0), outsize: 96
unique: 9, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 9, error: 0 (Undefined error: 0), outsize: 96
unique: 10, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 10, error: 0 (Undefined error: 0), outsize: 96
unique: 11, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 11, error: 0 (Undefined error: 0), outsize: 96
unique: 12, opcode: STATFS (17), nodeid: 1, insize: 40
unique: 3, opcode: STATFS (17), nodeid: 1, insize: 40
unique: 13, opcode: STATFS (17), nodeid: 1, insize: 40
unique: 14, opcode: STATFS (17), nodeid: 1, insize: 40
unique: 15, opcode: STATFS (17), nodeid: 1, insize: 40
unique: 16, opcode: STATFS (17), nodeid: 1, insize: 40
unique: 2, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 12, error: 0 (Undefined error: 0), outsize: 96
   unique: 3, error: 0 (Undefined error: 0), outsize: 96
   unique: 13, error: 0 (Undefined error: 0), outsize: 96
   unique: 14, error: 0 (Undefined error: 0), outsize: 96
   unique: 15, error: 0 (Undefined error: 0), outsize: 96
   unique: 16, error: 0 (Undefined error: 0), outsize: 96
   unique: 2, error: 0 (Undefined error: 0), outsize: 96
unique: 4, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 4, error: 0 (Undefined error: 0), outsize: 96
unique: 5, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 5, error: 0 (Undefined error: 0), outsize: 96
unique: 6, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 6, error: 0 (Undefined error: 0), outsize: 96
unique: 8, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 8, error: 0 (Undefined error: 0), outsize: 96
unique: 9, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 9, error: 0 (Undefined error: 0), outsize: 96
unique: 10, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 10, error: 0 (Undefined error: 0), outsize: 96
unique: 11, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 11, error: 0 (Undefined error: 0), outsize: 96
unique: 7, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 7, error: 0 (Undefined error: 0), outsize: 96
unique: 12, opcode: STATFS (17), nodeid: 1, insize: 40
   unique: 12, error: 0 (Undefined error: 0), outsize: 96
^Cunique: 3, opcode: DESTROY (38), nodeid: 1, insize: 40
   unique: 3, error: 0 (Undefined error: 0), outsize: 16
Apr 13 14:23:11  rar2fs[48662] <Debug>: unmounted /Users/user/Downloads/test/mnt
```

I've built rar2fs using unrar 5.9.2 and I've tried osxfuse 3.10.4 (Latest) as well as 3.8.3 but none of them worked. 

```
○ rar2fs --version                                                                                                      
rar2fs v1.29.0 (DLL version 8)    Copyright (C) 2009 Hans Beckerus
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it under
certain conditions; see <http://www.gnu.org/licenses/> for details.
MacFUSE library version: FUSE 2.7.5 / OSXFUSE 3.8.3
no mount point
```

**Comment by richardsimko** [comment]
Reverting to 1.26.0 solves the issue even with the latest osxfuse so it seems to be an issue with rar2fs

**Comment by hasse69** [comment]
Thanks for the issue reports.
Unfortunately I have very limited possibilities to try this myself and last time I checked rar2fs worked fine on OS X. The delta between latest release (1.29.0) and 1.26.0 is huge. I think you need to narrow it down somewhat to tell in what exact version things started to fail for you.

**Comment by hasse69** [comment]
Also note that 1.26.0 does not link statically to libunrar which suggest you probably linked to some old version unless you rebuilt libunrar and completely re-installed it. Standing in the repo/source root folder simply doing 
``´
ldd src/rar2fs
```
or 
```
src/rar2fs --version
```
would probably tell

**Comment by richardsimko** [comment]
I rebuilt libunrar for both the installs, as per the install instructions. With 1.29.0 I used 5.9.2 (Which is the latest one) and with 1.26.0 I used 5.4.5 as 5.9.2 was not compatible. For now those are the only two combinations I've tried.

I'm happy to test other combinations for you if you know which ones are worth trying, or versions of rar2fs between 1.26.0 and 1.29.0 if you would like. :)

**Comment by richardsimko** [comment]
Trying to build 1.29.0 against 5.4.5 of libunrar yields the following error, not sure if it's helpful:

```
mv -f .deps/rar2fs.Tpo .deps/rar2fs.Po
g++ -o rar2fs dllext.o optdb.o filecache.o iobuffer.o sighandler.o hashtable.o dircache.o dirlist.o rarconfig.o dirname.o rar2fs.o -D_THREAD_SAFE -pthread -rdynamic    -lfuse /Users/user/Downloads/rar2fs-1.29.0/unrar/libunrar.a -liconv -ldl
Undefined symbols for architecture x86_64:
  "RarTime::GetRaw()", referenced from:
      _RARListArchiveEx in dllext.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [rar2fs] Error 1
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
```

**Comment by hasse69** [comment]
Ok, if you are sure you have built everything properly then I guess it is not very likely to be a user error. Make sure that when using later version of libunrar that you link statically (which is the default if ./configure can locate `libunrar.a`). You can double check if it linked statically by using `ldd` on the compiled rar2fs binary. Again, I have had no recent reports of rar2fs not working on OS X and last time I checked it worked fine. I cannot tell exactly what version that was used back then but it was definitely while working with issue #119 so for sure something later than 1.26.0. 
I would try all versions one by one from 1.26.0 to 1.29.0.

**Comment by hasse69** [comment]
> Trying to build 1.29.0 against 5.4.5 of libunrar yields the following error, not sure if it's helpful:
> 
> ```
> mv -f .deps/rar2fs.Tpo .deps/rar2fs.Po
> g++ -o rar2fs dllext.o optdb.o filecache.o iobuffer.o sighandler.o hashtable.o dircache.o dirlist.o rarconfig.o dirname.o rar2fs.o -D_THREAD_SAFE -pthread -rdynamic    -lfuse /Users/user/Downloads/rar2fs-1.29.0/unrar/libunrar.a -liconv -ldl
> Undefined symbols for architecture x86_64:
>   "RarTime::GetRaw()", referenced from:
>       _RARListArchiveEx in dllext.o
> ld: symbol(s) not found for architecture x86_64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [rar2fs] Error 1
> make[1]: *** [all] Error 2
> make: *** [all-recursive] Error 1
> ```

I will need to check this one. We are supposed to support most pre-historic versions of libunrar.

**Comment by richardsimko** [comment]
FYI macOS doesn't have `ldd` but according to Google the equivalent is `otool -L` which when I run it on `src/rar2fs` in version 1.29.0 I get the following output:

```
src/rar2fs:
	/usr/local/lib/libfuse.2.dylib (compatibility version 10.0.0, current version 10.5.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
```

And on the 1.26.0 I see:
```
rar2fs:
	/usr/local/lib/libfuse_ino64.2.dylib (compatibility version 10.0.0, current version 10.5.0)
	libunrar.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

So there's definitely a difference there, not sure if it's a hint to anything though.

**Comment by richardsimko** [comment]
I tried 1.27.0, 1.27.2, and 1.28.0, all of them with 5.4.5 of libunrar. It seems to break at 1.28.0, prior versions work. They all have the same output from `otool -L` as 1.29.0 though.

**Comment by hasse69** [comment]
I still suspect there is something weird going on here with your build and that there is a version mismatch with what you build and link against, which must be of the same version.
I found the root cause of the failed symbol lookup, and it was caused by a bug in configure.ac that did not consider that source code is now built from the src folder. That made linking pick the *wrong* library from my system path rather then the one I just built in my local unrar folder.

**Comment by hasse69** [comment]
It is not possible that `otool` gives you the same output as 1.29.0 for any version of rar2fs built against libunrar 5.4.5 because of the simple reason that version does not create a static library. So please try to avoid using 5.4.5 completely and use something more recent and then try 1,28,0 again.

**Comment by richardsimko** [comment]
Sorry, I mistyped! I used libunrar 5.9.2 in all cases above!

**Comment by hasse69** [comment]
This is however a bit weird
```
	/usr/local/lib/libfuse.2.dylib (compatibility version 10.0.0, current version 10.5.0)

```
vs
```
	/usr/local/lib/libfuse_ino64.2.dylib (compatibility version 10.0.0, current version 10.5.0)
```
Are these really the same files? I do not think so. And why is it picking a different file for 1.26,0?
Is OSXFUSE really providing both of these libraries? I wonder if the library search order that I changed some time back is related. Perhaps that was done starting from 1.28.0. Can you check what version of fuse is picked for 1.27.0 and 1.27.2 please.

**Comment by richardsimko** [comment]
1.27.2:
```
rar2fs:
	/usr/local/lib/libfuse_ino64.2.dylib (compatibility version 10.0.0, current version 10.5.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

1.27.0:
```
rar2fs:
	/usr/local/lib/libfuse_ino64.2.dylib (compatibility version 10.0.0, current version 10.5.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
```

**Comment by hasse69** [comment]
Great! Ok, then I think we might have found the issue. I am still a bit confused though to why this has never been spotted before. Most likely it is your architecture that is very different from what I have been able to test on. Let's see if it is as simple as this. 

In the configure.ac file you should be able to find a few line looking like this
```
AC_CHECK_LIB(fuse,fuse_new, [FUSE_LIBS="-lfuse"],
    [ AC_CHECK_LIB(dokanfuse1.dll,fuse_new, [FUSE_LIBS="-ldokanfuse1.dll"],
        [ AC_CHECK_LIB(osxfuse,fuse_new, [FUSE_LIBS="-losxfuse"],
            [ AC_CHECK_LIB(fuse4x,fuse_new, [FUSE_LIBS="-lfuse4x"],
                [ AC_CHECK_LIB(fuse_ino64,fuse_new, [FUSE_LIBS="-lfuse_ino64"],
```
Now change these too
```
AC_CHECK_LIB(fuse_ino64,fuse_new, [FUSE_LIBS="-lfuse_ino64"],
    [ AC_CHECK_LIB(dokanfuse1.dll,fuse_new, [FUSE_LIBS="-ldokanfuse1.dll"],
        [ AC_CHECK_LIB(osxfuse,fuse_new, [FUSE_LIBS="-losxfuse"],
            [ AC_CHECK_LIB(fuse4x,fuse_new, [FUSE_LIBS="-lfuse4x"],
                [ AC_CHECK_LIB(fuse,fuse_new, [FUSE_LIBS="-lfuse"],
```
Basically swapping search for `libfuse_ino64` with `libfuse`.
After the change, re-run configure and make sure it find the proper library, e.g.
```
checking for fuse_new in -lfuse_ino64... yes
```

**Comment by richardsimko** [comment]
That did the trick! Now 1.29.0 works! Thanks for the stellar help! And thanks for a great piece of software, it's served me well for many years now 🎉 

I experimented a bit with fuse as well and realized that the macfuse compatibility layer must be installed in order for rar2fs to work, just a side note that might be worth mentioning in the install instructions, otherwise it fails when looking for `fuse..h`.

**Comment by hasse69** [comment]
Are you sure about that? I think the MacFUSE layer was only there to provide support for some very ancients systems? So you are saying that OSXFUSE does not install the header files? That seems a bit odd to be honest or OSXFUSE does by default support target development but only run-time in that case.

I will think about how to solve your issue in a proper patch, probably as simple as doing what we just did. But I need a couple of iterations in my head before I can tell for sure.

**Comment by richardsimko** [comment]
If I run it without that installed I get the following output from `./configure`:
```
checking fuse.h usability... no
checking fuse.h presence... no
checking for fuse.h... no
configure: error:
        Cannot find fuse.h - add alternative include search path using
        --with-fuse and rerun configure.
```

I also only have the following fuse libs:
```
-rwxr-xr-x    1 root  wheel   951K  5 Dec 13:32 libosxfuse.2.dylib
lrwxr-xr-x    1 root  wheel    18B 13 Apr 22:06 libosxfuse.dylib -> libosxfuse.2.dylib
-rwxr-xr-x    1 root  wheel   933B  5 Dec 13:32 libosxfuse.la
lrwxr-xr-x    1 root  wheel    18B 13 Apr 22:06 libosxfuse_i64.2.dylib -> libosxfuse.2.dylib
lrwxr-xr-x    1 root  wheel    16B 13 Apr 22:06 libosxfuse_i64.dylib -> libosxfuse.dylib
lrwxr-xr-x    1 root  wheel    13B 13 Apr 22:06 libosxfuse_i64.la -> libosxfuse.la
```

With compatibility installed I get these:
```
lrwxr-xr-x    1 root  wheel    13B 13 Apr 22:11 libfuse.0.dylib -> libfuse.dylib
-rwxr-xr-x    1 root  wheel   615K  5 Dec 13:33 libfuse.2.dylib
lrwxr-xr-x    1 root  wheel    15B 13 Apr 22:11 libfuse.dylib -> libfuse.2.dylib
-rwxr-xr-x    1 root  wheel   917B  5 Dec 13:33 libfuse.la
-rwxr-xr-x    1 root  wheel   615K  5 Dec 13:33 libfuse_ino64.2.dylib
lrwxr-xr-x    1 root  wheel    21B 13 Apr 22:11 libfuse_ino64.dylib -> libfuse_ino64.2.dylib
-rwxr-xr-x    1 root  wheel   947B  5 Dec 13:33 libfuse_ino64.la
-rwxr-xr-x    1 root  wheel   951K  5 Dec 13:32 libosxfuse.2.dylib
lrwxr-xr-x    1 root  wheel    18B 13 Apr 22:11 libosxfuse.dylib -> libosxfuse.2.dylib
-rwxr-xr-x    1 root  wheel   933B  5 Dec 13:32 libosxfuse.la
lrwxr-xr-x    1 root  wheel    18B 13 Apr 22:11 libosxfuse_i64.2.dylib -> libosxfuse.2.dylib
lrwxr-xr-x    1 root  wheel    16B 13 Apr 22:11 libosxfuse_i64.dylib -> libosxfuse.dylib
lrwxr-xr-x    1 root  wheel    13B 13 Apr 22:11 libosxfuse_i64.la -> libosxfuse.la
```

So without compatibility the `libfuse_ino64.2.dylib` is missing.

**Comment by hasse69** [comment]
Interesting, but I think what you really should use here is libosxfuse_i64?

**Comment by richardsimko** [comment]
I'm not the one selecting ;)

But it seems strange to me as well that it doesn't work, since libosxfuse is in the list that it's searching for? I'm presuming it's this line?

```
        [ AC_CHECK_LIB(osxfuse,fuse_new, [FUSE_LIBS="-losxfuse"],
```

**Comment by hasse69** [comment]
Yea, but it is a exclusive search order, right, So if it finds something before that line it will pick that one.
Can you play a little bit with this and replace the first line with
```
AC_CHECK_LIB(osxfuse_i64,fuse_new, [FUSE_LIBS="-losxfuse_i64"],
```
rather than
```
AC_CHECK_LIB(fuse,fuse_new, [FUSE_LIBS="-lfuse"],
```
and see what happens?

**Comment by hasse69** [comment]
So, maybe the order in which to search for libraries should really be

osxfuse_i64
osxfuse
fuse_ino64
fuse4x
fuse
dokanfuse1

This would effectively mean Linux users etc, would need to fail a few times before finding its fuse library. But we need to bring the "odd" birds to the front I guess. Another more complicated solution would be to check specifically for each OS and make decisions based on that. That is however something I would like to avoid though since if you ask me such tests are error prone.

**Comment by richardsimko** [comment]
Of course, but if I uninstall the compatibility layer I though it would go down and get to the `libosxfuse` line.

I tried with the new order and it worked well, the drive even gets a nice new icon as well!

<img width="74" alt="image" src="https://user-images.githubusercontent.com/1245031/79160778-a69ef800-7dda-11ea-88d6-c048bbceed75.png">

**Comment by hasse69** [comment]
Yes, but if you drop the compatibility layer you lack the proper header files?

**Comment by richardsimko** [comment]
Ahh, then it makes sense!

However the new order you suggested (Putting `AC_CHECK_LIB(osxfuse_i64,fuse_new, [FUSE_LIBS="-losxfuse_i64"],` at the top) still doesn't work without the compatibility layer, still missing the headers. But perhaps that was expected?

**Comment by hasse69** [comment]
It was expected judging from your findings, yes.
What we had to check was that mixing the headers from the compatibility layer and the "real" OSXFUSE library still works. And if I interpret your tests correctly it in fact did?

**Comment by richardsimko** [comment]
Yep! Granted I haven't tried every aspect of rar2fs but mounting a basic test RAR-file worked well.

**Comment by hasse69** [comment]
Ok, please try some more "advanced" operations on your file system and then report back.
It is vital we understand the effect of mixing these two before we can safely change the search order according to my last suggestion.

**Comment by hasse69** [comment]
And honestly I think you should post a question on the OSXFUSE forum to get a better understanding of why only the compatibility layer installs the required header files for any FUSE based file system.

**Comment by richardsimko** [comment]
I tried mounting multi-part RARs as well which works, not sure what else I should test for as that's all I use it for. I can run this for a few days and report back if I encounter any issues.

**Comment by hasse69** [comment]
Please do that. I will prepare a patch for this but will put it on hold until you report back your findings.
Thanks a lot for you patience and assistance in testing this.

**Comment by richardsimko** [comment]
Sounds good! Thank you for all the help :)

**Comment by hasse69** [comment]
Btw, I checked the issue tracker on OSXFUSE and according to that there should be header files installed in `/usr/local/include/osxfuse`. Can you check if that is the case on your system?

**Comment by richardsimko** [comment]
Hmm, yeah there is indeed a `fuse.h` there even if I install it without the compatibility layer.

**Comment by hasse69** [comment]
Ah, ok. Then what you should try is to configure like this
```
./configure --with-fuse=/usr/local/include/osxfuse
```
Let's see if that works now without the compatibility layer installed.

**Comment by richardsimko** [comment]
That worked perfectly, now built and running without the compatibility layer 🎉

**Comment by hasse69** [comment]
Great! Then I will await your feedback to make sure everything runs smoothly for a couple of days before submitting the patch.

**Comment by hasse69** [comment]
Any updates here?

**Comment by richardsimko** [comment]
Still working well! I use it mostly for multipart archives. Notifications about file changes are also working as expected (Which it wasn't before) so I'm happy 🎉

**Comment by hasse69** [comment]
Thanks for the feedback. Patch is now merged into master.

**Comment by richardsimko** [comment]
Awesome, thanks for all your help and the great software!
